File: 0061-rotate-list/0061-rotate-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode rotateRight(ListNode head, int k) {
        if( head == null || head.next == null || k==0) return head; 
        int l = 1 ; 
        ListNode temp = head; 
        while( temp.next != null){
            l+=1; 
            temp = temp.next; 
        }
        temp.next = head;

        k = k%l; 
        int a = l-k; 
        ListNode temp2 = head;
        for( int i =0 ; i< a-1 ; i++){
                temp2 = temp2.next;
        }
        ListNode temp3 = temp2.next; 
        temp2.next = null; 
        return temp3;

        
    }
}

File: 0061-rotate-list/0061-rotate-list.cpp
class Solution {
public:
    ListNode* rotateRight(ListNode* head, int k) {
        if (!head || !head->next || k == 0) return head;

        // Step 1: Count the length
        int len = 1;
        ListNode* tail = head;
        while (tail->next) {
            tail = tail->next;
            len++;
        }

        // Step 2: Make it circular
        tail->next = head;

        // Step 3: Break at new tail
        k = k % len;
        int stepsToNewTail = len - k;
        ListNode* newTail = head;
        for (int i = 1; i < stepsToNewTail; i++) {
            newTail = newTail->next;
        }

        ListNode* newHead = newTail->next;
        newTail->next = nullptr;

        return newHead;
    }
};
